base/reduce.jl

```diff
@@ -25,8 +31,14 @@ The reduction operator used in `prod`. The main difference from [`*`](@ref) is t
 integers are promoted to `Int`/`UInt`.
 """
 mul_prod(x, y) = x * y
-mul_prod(x::BitSignedSmall, y::BitSignedSmall) = Int(x) * Int(y)
-mul_prod(x::BitUnsignedSmall, y::BitUnsignedSmall) = UInt(x) * UInt(y)
+mul_prod(x::BitSigned, y::BitSigned) = is_small_int_type(x) && is_small_int_type(y) ?
+    Int(x) * Int(y) :
```

Comment:
```suggestion
    Int(x)::Int * Int(y)::Int :
```

Reply:
Unfortunately, it doesn't seem to help; beside your suggestion, I tried a `::Real` annotation, and also as a function return type annotation.

Reply:
Both should be necessary.

Reply:
Have you tried both at the same time?

Reply:
I had tried with all `::Real` annotations I could think of, and now with:
```
 add_sum(x::BitSigned, y::BitSigned)::Real = is_small_int_type(x) && is_small_int_type(y) ?
    Int(x)::Int + Int(y)::Int :
    (x + y)::Real

add_sum(x::BitUnsigned, y::BitUnsigned)::Real = is_small_int_type(x) && is_small_int_type(y) ?
    UInt(x)::UInt + UInt(y)::UInt :
    (x + y)::Real
```
It still doesn't work.